### ISSUE TITLE
Better understanding of each parameter

### ISSUE THREAD

**Issue by Juanjojr9** [ultralytics/yolov5#6507]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello.

I tried to look for information about different parameters that can be used in train.py, val.py and detect.py, but I could not find information on some. I have tested activating each parameter, but I can't understand what they are for.
Is there an article explaining what each is for?

I would like to understand each of them as well as possible and make the most of them. I am very interested in this topic

I do not know what these parameters are used for, i.e. what their functions are, nor do I know when I should use them. Also I have other questions about some of them

In the train.py file:

- What is --evolve for and how is it used?
- --bucket
- What is the logic of using --multi-scale, if the images change size when using --hyp? Since one hyperparameter is the scale: 0.5 
- --workers
- --linear-lr
- --local_rank
- --label-smoothing
Why if I put --nosave, it doesn't do the best.pt validation, nor does it do any metrics, like the PR curve?
- --quad

In the val.py file:

- --task (speed and study)
- --augment
- --half
- --dnn

In the detect.py file:

-  --agnostic-nms
-  --update

I hope you can help me to solve this doubt.

Thank you very much for your help.

JJ

**Comment by glenn-jocher** [comment]
@Juanjojr9 we are working on improved documentation, but for the time being you can see the help field in each argument for details. If you have doubts about arguments, always use the default value.
https://github.com/ultralytics/yolov5/blob/4c409332667477560200958b513b958bb8fdef71/train.py#L453-L496

For --evolve see **Hyperparameter Evolution** tutorial below:

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [TFLite, ONNX, CoreML, TensorRT Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

Good luck 🍀 and let us know if you have any other questions!

**Comment by Juanjojr9** [comment]
Hello @glenn-jocher,
First of all, I would like to thank you for your response.

Currently, I'm doing my final project for graduation using YOLOv5. That's why I'm so interested in learning as much as I can, and in case I am asked by my tutor, to be able to answer in a coherent manner.
I have read the help field in each argument, but it is not very detailed.

This is why I would like to know what each parameter does in a short form and I can defend myself 

Any help is good.

JJ

**Comment by glenn-jocher** [comment]
@Juanjojr9 thanks for your feedback! We will try to improve the documentation going forward to explain arguments in greater detail. For now the help section is all that is available, and as I said before if you are unsure of an argument then there is no reason to modify it.

**Comment by glenn-jocher** [comment]
@Juanjojr9 also I've added our tips for best training here which explain a few concepts.

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by Tautvydas-byte** [comment]
Hello, could you explain what is _translate: 0.1  # image translation (+/- fraction)_ and _perspective: 0.0  # image perspective (+/- fraction), range 0-0.001_. Does image translate means something what is written in this article Image-to-Image Translation: Methods and Applications https://arxiv.org/pdf/2101.08629.pdf and image perspective is something like https://www.geeksforgeeks.org/perspective-transformation-python-opencv/ ?

**Comment by glenn-jocher** [comment]
@Tautvydas-byte translation left-right or up-down

**Comment by Reynolddoss** [comment]
@glenn-jocher How would you suggest the retraining in the below case?
I have trained a model with 30 classes but I had the training data for just 25 classes.
Now I have the training data for 5 classes that were not included before. How can I retrain for this using the previously trained model? without using the previous data

**Comment by bit-scientist** [comment]
@Reynolddoss  you must have your previous dataset whether you add or subtract from the first used dataset. You are dealing with the weights, they are updated in each iteration using gradient decent, so if you don't have a dataset of which you are planning to detect objects from, then there is no object in the first place.
